// react-noop-renderer/cjs/react-noop-renderer-flight-server.development.js
'use strict';

var flightModules = require('react-noop-renderer/flight-modules');
var ReactFlightServer = require('react-server/flight');

/**
 * This is a renderer of React that doesn't have a render target output.
 * It is useful to demonstrate the internals of the reconciler in isolation
 * and for testing semantics of reconciliation separate from the host
 * environment.
 */
var ReactNoopFlightServer = ReactFlightServer({
  scheduleWork: function (callback) {
    callback();
  },
  beginWriting: function (destination) {},
  writeChunk: function (destination, chunk) {
    destination.push(chunk);
  },
  writeChunkAndReturn: function (destination, chunk) {
    destination.push(chunk);
    return true;
  },
  completeWriting: function (destination) {},
  close: function (destination) {},
  closeWithError: function (destination, error) {},
  flushBuffered: function (destination) {},
  stringToChunk: function (content) {
    return content;
  },
  stringToPrecomputedChunk: function (content) {
    return content;
  },
  isModuleReference: function (reference) {
    return reference.$$typeof === Symbol.for('react.module.reference');
  },
  getModuleKey: function (reference) {
    return reference;
  },
  resolveModuleMetaData: function (config, reference) {
    return flightModules.saveModule(reference.value);
  }
});

function render(model, options) {
  var destination = [];
  var bundlerConfig = undefined;
  var request = ReactNoopFlightServer.createRequest(model, bundlerConfig, options ? options.onError : undefined, options ? options.context : undefined, options ? options.identifierPrefix : undefined);
  ReactNoopFlightServer.startWork(request);
  ReactNoopFlightServer.startFlowing(request, destination);
  return destination;
}

exports.render = render;
