// react-suspense-test-utils/cjs/react-suspense-test-utils.js
'use strict';

var React = require('react');

var ReactSharedInternals = React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;

var ReactCurrentDispatcher = ReactSharedInternals.ReactCurrentDispatcher;

function unsupported() {
  throw new Error('This feature is not supported by ReactSuspenseTestUtils.');
}

function waitForSuspense(fn) {
  var cache = new Map();
  var testDispatcher = {
    getCacheForType: function (resourceType) {
      var entry = cache.get(resourceType);

      if (entry === undefined) {
        entry = resourceType(); // TODO: Warn if undefined?

        cache.set(resourceType, entry);
      }

      return entry;
    },
    readContext: unsupported,
    useContext: unsupported,
    useMemo: unsupported,
    useReducer: unsupported,
    useRef: unsupported,
    useState: unsupported,
    useInsertionEffect: unsupported,
    useLayoutEffect: unsupported,
    useCallback: unsupported,
    useImperativeHandle: unsupported,
    useEffect: unsupported,
    useDebugValue: unsupported,
    useDeferredValue: unsupported,
    useTransition: unsupported,
    useId: unsupported,
    useMutableSource: unsupported,
    useSyncExternalStore: unsupported,
    useCacheRefresh: unsupported
  }; // Not using async/await because we don't compile it.

  return new Promise(function (resolve, reject) {
    function retry() {
      var prevDispatcher = ReactCurrentDispatcher.current;
      ReactCurrentDispatcher.current = testDispatcher;

      try {
        var result = fn();
        resolve(result);
      } catch (thrownValue) {
        if (typeof thrownValue.then === 'function') {
          thrownValue.then(retry, retry);
        } else {
          reject(thrownValue);
        }
      } finally {
        ReactCurrentDispatcher.current = prevDispatcher;
      }
    }

    retry();
  });
}

exports.waitForSuspense = waitForSuspense;
